src/matrix_multiply.jl

```diff
@@ -1,676 +1,303 @@
-    else
-        exprs = [zero(T) for k = 1:sA[1]]
-    end
-
-    return quote
-        $(Expr(:meta,:inline))
-        if length(b) != $(sA[2])
-            error("Dimension mismatch")
-        end
-        @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
-    end
-end
-
-# Ambiguity with BLAS function
-@generated function *{TA <: Base.LinAlg.BlasFloat, Tb}(A::StaticMatrix{TA}, b::StridedVector{Tb})
-    sA = size(A)
-
-    s = Size(sA[1])
-    T = promote_matprod(TA, Tb)
-
-    if T == TA
-        newtype = similar_type(A, s)
-    else
-        newtype = similar_type(A, T, s)
-    end
-
-    if sA[2] != 0
-        exprs = [reduce((ex1,ex2) -> :(+($ex1,$ex2)), [:(A[$(sub2ind(sA, k, j))]*b[$j]) for j = 1:sA[2]]) for k = 1:sA[1]]
-    else
-        exprs = [zero(T) for k = 1:sA[1]]
-    end
-
-    return quote
-        $(Expr(:meta,:inline))
-        if length(b) != $(sA[2])
-            error("Dimension mismatch")
-        end
-        @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
-    end
-end
-
-@generated function *(a::StaticVector, B::StaticMatrix)
-    Ta = eltype(a)
-    TB = eltype(B)
-    sa = size(a)
-    sB = size(B)
-
-    s = Size(sa[1],sB[2])
-    T = promote_matprod(Ta, TB)
-
-    if sB[1] != 1
-        error("Dimension mismatch")
-    end
-
-    if s == sB
-        if T == TB
-            newtype = B
-        else
-            newtype = similar_type(B, T)
-        end
-    else
-        if T == TB
-            newtype = similar_type(B, s)
-        else
-            newtype = similar_type(B, T, s)
-        end
-    end
-
-    exprs = [:(a[$j] * B[$k]) for j = 1:s[1], k = 1:s[2]]
+# outer product
+@generated function _A_mul_B(::Size{sa}, ::Size{sb}, a::StaticVector{Ta}, b::RowVector{Tb, <:StaticVector}) where {sa, sb, Ta, Tb}
+    newsize = (sa[1], sb[2])
+    exprs = [:(a[$i]*b[$j]) for i = 1:sa[1], j = 1:sb[2]]
 
     return quote
-        $(Expr(:meta,:inline))
-        @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
+        @_inline_meta
+        T = promote_op(*, Ta, Tb)
+        @inbounds return similar_type(b, T, Size($newsize))(tuple($(exprs...)))
     end
 end
 
-#@inline *{S1,S2,S3}(A::MMatrix{S1,S2}, B::MMatrix{S2,S3}) = MMatrix{S1,S3}(SMatrix{S1,S2}(A)*SMatrix{S2,S3}(B))
+# TODO: I removed StaticMatrix * AbstractVector. Reinstate?
 
-@generated function *(A::StaticMatrix, B::StaticMatrix)
-    TA = eltype(A)
-    TB = eltype(B)
-
-    T = promote_matprod(TA, TB)
-
-    can_mutate = !isbits(A) || !isbits(B) # !isbits implies can get a persistent pointer (to pass to BLAS). Probably will change to !isimmutable in a future version of Julia.
-    can_blas = T == TA && T == TB && T <: Union{Float64, Float32, Complex{Float64}, Complex{Float32}}
+@generated function _A_mul_B(Sa::Size{sa}, Sb::Size{sb}, a::StaticMatrix{Ta}, b::StaticMatrix{Tb}) where {sa, sb, Ta, Tb}
+    can_mutate = a.mutable && b.mutable # TODO this probably isn't safe. Maybe a trait??
```

Comment:
It's a bit funny... you can have a `StaticArray` struct which is immutable but for which `setfield!` works (such as `SizedArray`), and you can (in theory) have a `mutable struct` and not define `setfield!`. Annoying...

I probably should just specialize on `Union{SizedMatrix{<:BlasFloat}, MMatrix{<:BlasFloat}}`.